bin/dump-to-s3-mysql.sh

```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+
+# exit on failures
+set -e
+set -o pipefail
+
+usage() {
+  echo "Usage: $(basename "$0") [OPTIONS]" 1>&2
+  echo "  -h                - help"
+  echo "  -b <bucket_name>  - S3 Bucket name"
+  echo "  This script makes an SQL dump of all databases within a given MySQL"
+  echo "  server, and uploads them to the specified S3 bucket"
+  echo "  The MySQL target and credentials must be specified as environment variables:"
+  echo "  'DB_HOST', 'DB_USER' and 'DB_PASSWORD'"
+  exit 1
+}
+
+while getopts "b:h" opt; do
+  case $opt in
+    b)
+      BUCKET_NAME=$OPTARG
+      ;;
+    h)
+      usage
+      ;;
+    *)
+      usage
+      ;;
+  esac
+done
+
+if [[
+  -z "$BUCKET_NAME"
+  || -z "$DB_HOST"
+  || -z "$DB_USER"
+  || -z "$DB_PASSWORD"
+]]
+then
+  usage
+fi
+
+echo "==> Starting backup of MySQL Server $DB_HOST ..."
+DATE_STRING="$(date +%Y%m%d%H%M)"
+DUMP_DIR="/tmp/sqlbackups/$DB_HOST"
+mkdir -p "$DUMP_DIR"
+
+echo "==> Getting database names ..."
+DATABASES="$(mysql -N \
+  -u "$DB_USERNAME" \
+  -p"$DB_PASSWORD" \
+  -h "$DB_HOST" \
+  -e 'show databases' \
+  | grep -Ev 'Databases|information_schema|performance_schema|sys'
+)"
+
+while IFS='' read -r DB_NAME
+do
+  DUMP_TARGET="/$DUMP_DIR/$DATE_STRING-$DB_NAME.sql"
+  echo "==> Dumping '$DB_NAME' database to $DUMP_TARGET ..."
+  mysqldump \
+    -u "$DB_USER" \
+    -p"$DB_PASSWORD" \
+    -h "$DB_HOST" \
+    --set-gtid-purged=OFF \
+    --column-statistics=0 \
+    --single-transaction "$DB_NAME" > "$DUMP_TARGET"
```

Comment:
If databases are quite large, a single transaction requires that all tables get locked during the transaction. This could mean sites stop working or hang whilst waiting for the lock to be freed. Maybe worth adding `--skip-lock-tables` too to bypass that?

Reply:
Afaik the locking of tables only adds a lock to Read (To allow the dump to run faster)
Whilst taking the dump, inserts are still locked regardless
I think we should at least have the lock options configurable - which we can add if needed 👍